doc/src/devdocs/EscapeAnalysis/README.md

```diff
@@ -0,0 +1,322 @@
+```@docs
+Base.code_escapes
+InteractiveUtils.@code_escapes
+```
+
+## Analysis Design
+
+### Lattice Design
+
+`EscapeAnalysis` is implemented as a [data-flow analysis](https://en.wikipedia.org/wiki/Data-flow_analysis)
+that works on a lattice called `x::EscapeLattice`, which is composed of the following properties:
+- `x.Analyzed::Bool`: not formally part of the lattice, only indicates `x` has not been analyzed or not
+- `x.ReturnEscape::BitSet`: records SSA statements where `x` can escape to the caller via return
+- `x.ThrownEscape::BitSet`: records SSA statements where `x` can be thrown as exception
+  (used for the [exception handling](@ref EA-Exception-Handling) described below)
+- `x.AliasInfo`: maintains all possible values that can be aliased to fields or array elements of `x`
+  (used for the [alias analysis](@ref EA-Alias-Analysis) described below)
+- `x.ArgEscape::Int` (not implemented yet): indicates it will escape to the caller through
+  `setfield!` on argument(s)
+
+These attributes can be combined to create a partial lattice that has a finite height, given
+the invariant that an input program has a finite number of statements, which is assured by Julia's semantics.
+The clever part of this lattice design is that it enables a simpler implementation of
+lattice operations by allowing them to handle each lattice property separately[^LatticeDesign].
+
+### Backward Escape Propagation
+
+This escape analysis implementation is based on the data-flow algorithm described in the paper[^MM02].
+The analysis works on the lattice of `EscapeLattice` and transitions lattice elements from the
+bottom to the top until every lattice element gets converged to a fixed point by maintaining
+a (conceptual) working set that contains program counters corresponding to remaining SSA
+statements to be analyzed. The analysis manages a single global state that tracks
+`EscapeLattice` of each argument and SSA statement, but also note that some flow-sensitivity
+is encoded as program counters recorded in `EscapeLattice`'s `ReturnEscape` property,
+which can be combined with domination analysis later to reason about flow-sensitivity if necessary.
+
+One distinctive design of this escape analysis is that it is fully _backward_,
+i.e. escape information flows _from usages to definitions_.
+For example, in the code snippet below, EA first analyzes the statement `return %1` and
+imposes `ReturnEscape` on `%1` (corresponding to `obj`), and then it analyzes
+`%1 = %new(Base.RefValue{String, _2}))` and propagates the `ReturnEscape` imposed on `%1`
+to the call argument `_2` (corresponding to `s`):
+```julia
+julia> code_escapes((String,)) do s
+           obj = Ref(s)
+           return obj
+       end
+#1(↑ _2::String) in Main at REPL[2]:2
+2 ↑  1 ─ %1 = %new(Base.RefValue{String}, _2)::Base.RefValue{String}                │╻╷╷ Ref
+3 ◌  └──      return %1                                                             │
```

Comment:
can you turn off the function inlining info on the right? it is rather distracting and not good at line wrapping

Reply:
Do you want it to be turned off globally (including REPL usages for debugging), or just want it not to be included in documentations?

Reply:
Globally is probably okay too. It makes it hard to copy-paste currently.

Reply:
02c5fd5